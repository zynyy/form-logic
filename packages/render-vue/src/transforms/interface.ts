import { BtnFieldsItem, LogicListItem, MetaSchema, MetaSchemaData, SchemaPattern } from '@/interface';
import { ISchema } from '@formily/json-schema';

export interface TransformsSchemaOptions {
  metaSchema: MetaSchema;
  pattern?: SchemaPattern;
  hasGroup?: boolean;
}

export interface ListSchema {
  searchSchema: ISchema;
  tableSchema: ISchema;
  hasCollapsed: boolean;
  searchLogic: LogicListItem[];
  tableLogic: LogicListItem[];
  searchBtnFields: BtnFieldsItem[];
  tableBtnFields: BtnFieldsItem[];
  searchButtons: MetaSchemaData[];
  transformsDone: boolean;
}

export interface FormSchema {
  schema: ISchema;
  btnSchema: ISchema;
  buttons: MetaSchemaData[];
  logicList: LogicListItem[];
  btnFields: BtnFieldsItem[];
  pattern: SchemaPattern;
  transformsDone: boolean;
}

export interface Properties {
  [code: string]: ISchema;
}

export interface PropertiesSchema {
  propertiesSchema: Properties;
  buttons: MetaSchemaData[];
  logicList: LogicListItem[];
  btnFields: BtnFieldsItem[];
  pattern: SchemaPattern;
  transformsDone: boolean;
}

export enum RunSchemaEnum {
  formSchema = 'formSchema',
  propertiesSchema = 'propertiesSchema',
  listSchema = 'listSchema',
}
