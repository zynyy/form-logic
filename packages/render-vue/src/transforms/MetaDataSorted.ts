import { MetaSchema, MetaSchemaData, MetaSchemaGroup } from '@/interface';

interface ResultReturn {
  columns: MetaSchemaData[];
  searchColumns: MetaSchemaData[];
  tableColumns: MetaSchemaData[];
  button: MetaSchemaData[];
  searchButton: MetaSchemaData[];
  tableButton: MetaSchemaData[];
}

class MetaDataSorted {
  private readonly metaSchema: MetaSchema;
  private readonly hasGroup: boolean = false;

  columnsArray: MetaSchemaData[] = [];

  searchColumnsArray: MetaSchemaData[] = [];

  tableColumnsArray: MetaSchemaData[] = [];

  buttonsArray: MetaSchemaData[] = [];

  searchButtonsArray: MetaSchemaData[] = [];

  tableButtonsArray: MetaSchemaData[] = [];

  groupsOrder: MetaSchemaGroup[] = [];

  groupButtonsArray: MetaSchemaData[] = [];

  constructor(metaSchema: MetaSchema, hasGroup?: boolean) {
    this.metaSchema = metaSchema;
    this.hasGroup = hasGroup;
  }

  private column = (item: MetaSchemaData) => {
    this.columnsArray.push(item);
  };

  private search_column = (item: MetaSchemaData) => {
    this.searchColumnsArray.push(item);
  };

  private table_column = (item: MetaSchemaData) => {
    this.tableColumnsArray.push(item);
  };

  private button = (item: MetaSchemaData) => {
    const { group } = item;

    if (this.hasGroup && group) {
      this.groupButtonsArray.push(item);
    } else {
      this.buttonsArray.push(item);
    }
  };

  private search_button = (item: MetaSchemaData) => {
    this.searchButtonsArray.push(item);
  };

  private table_button = (item: MetaSchemaData) => {
    this.tableButtonsArray.push(item);
  };

  run = () => {
    const { data, group } = this.metaSchema || {};

    data?.forEach((item) => {
      const { type } = item || {};
      this[type]?.(item);
    });

    const occupyAllCodes =
      group
        ?.filter((item) => {
          const { mode } = item || {};
          return !!mode;
        })
        ?.reduce((acc, cur) => {
          const { modeCodes, code } = cur || {};
          return acc.concat(code).concat(modeCodes);
        }, [])
        .filter((val) => val) || [];

    this.groupsOrder =
      group?.filter((item) => {
        const { mode, code } = item || {};
        return !occupyAllCodes.includes(code) || !!mode;
      }) || [];
  };

  result = (): ResultReturn => {
    this.run();
    return {
      columns: this.columnsArray,
      searchColumns: this.searchColumnsArray,
      tableColumns: this.tableColumnsArray,
      button: this.buttonsArray,
      searchButton: this.searchButtonsArray,
      tableButton: this.tableButtonsArray,
    };
  };
}

export default MetaDataSorted;
