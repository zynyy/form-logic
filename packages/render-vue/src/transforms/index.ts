export * from './interface';

export { default as TransformsSchema } from './TransformsSchema';

