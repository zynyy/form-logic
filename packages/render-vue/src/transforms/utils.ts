import { StrNumBool } from '@/interface';

export const strNumBoolToBoolean = (val?: StrNumBool) => {
  return val === undefined ? false : Boolean(Number(val));
};
