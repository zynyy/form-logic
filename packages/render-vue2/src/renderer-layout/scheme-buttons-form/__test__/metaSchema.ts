import { MetaSchema } from '@/interface';

export default {
  code: 'Api_C',
  labelCol: 6,
  wrapperCol: 18,
  defaultSearchColumn: 6,
  group: [],
  data: [
    {
      type: 'column',
      schemaType: 'string',
      required: '0',
      hidden: '0',
      disabled: '0',
      hiddenName: '0',
      wrap: '0',
      logics: [],
      code: 'url',
      name: '请求接口',
      hasSerialNo: '0',
      hasSort: '0',
    },
    {
      type: 'column',
      schemaType: 'string',
      required: '1',
      hidden: '0',
      disabled: '0',
      hiddenName: '0',
      wrap: '0',
      logics: [],
      code: 'method',
      name: '请求方式',
      defaultValue: 'get',
      component: 'input',
    },
  ],
  model: 'Api',
  codeSuffix: 'C',
  name: '组件接口配置',
} as MetaSchema;
