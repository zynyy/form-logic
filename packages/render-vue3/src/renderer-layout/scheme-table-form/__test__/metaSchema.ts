import { MetaSchema } from '@/interface';

export default {
  code: 'Logic_L',
  name: '逻辑列表',
  model: 'Logic',
  labelCol: 6,
  wrapperCol: 18,
  columnLayout: 3,
  group: [],
  defaultSearchColumn: 6,
  data: [
    {
      code: 'code',
      name: '编码',
      type: 'search_column',
      schemaType: 'string',
      logics: [],
    },
    {
      code: 'name',
      name: '名称',
      type: 'search_column',
      schemaType: 'string',
      logics: [],
    },
    {
      code: 'addBtn',
      name: '新建',
      type: 'button',
      schemaType: 'void',
      component: 'ArrayBase.Addition',
      logics: [],
    },
    {
      code: 'code',
      name: '编码',
      type: 'table_column',
      schemaType: 'string',
      logics: [],
    },
    {
      code: 'name',
      name: '名称',
      type: 'table_column',
      schemaType: 'string',
      logics: [],
    },
    {
      code: 'remarks',
      name: '备注说明',
      type: 'table_column',
      schemaType: 'string',
      logics: [],
    },
    {
      code: 'editBtn',
      name: '编辑',
      type: 'table_button',
      schemaType: 'void',
      component: 'ArrayBase.Edit',
      logics: [],
    },
    {
      code: 'removeBtn',
      name: '删除',
      type: 'table_button',
      schemaType: 'void',
      component: 'ArrayBase.Remove',
      logics: [],
    },
    {
      type: 'table_button',
      schemaType: 'string',
      required: '0',
      hidden: '0',
      disabled: '0',
      logics: [],
      code: 'detailBtn',
      name: '详情',
      component: 'ArrayBase.Detail',
    },
  ],
} as MetaSchema;
