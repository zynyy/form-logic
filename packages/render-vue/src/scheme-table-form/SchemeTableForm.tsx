import { setValidateLanguage } from '@formily/core';

import { FormProvider } from '@/formily-vue';
import { computed, defineComponent, ref, toRef, onMounted, watchEffect, watch } from 'vue';
import { Empty, Skeleton } from 'ant-design-vue';

import { useNotifyDevtools } from '@/hooks';
import useCreateSchemaField from '@/hooks/useSchemaField';

import { SchemeTableFormProps, getSchemeTableFormProps } from './interface';

import useDOMRect from '@/hooks/dom/useDOMRect';
import { Key } from 'ant-design-vue/lib/_util/type';
import { useItemKey } from '@/components/array-base';
import { toArray } from '@/utils';
import { LIST_FILED_CODE } from '@/utils/constant';

const SchemeTableForm = defineComponent({
  name: 'SchemeTableForm',
  inheritAttrs: false,
  props: getSchemeTableFormProps(),
  setup(props: SchemeTableFormProps) {
    const { components } = props;

    const dataSourceRef = toRef(props, 'dataSource');
    const formRef = toRef(props, 'form');
    const defaultSelectedRowsRef = toRef(props, 'selectedRows');

    const SchemaField = useCreateSchemaField();

    const [contentRect, domRef] = useDOMRect<HTMLDivElement>();

    useNotifyDevtools(toRef(props, 'form'));

    const selectedRowKeys = ref<Key[]>([]);

    const selectedRows = ref<any[]>([]);

    const setSelectedRows = (nextSelectedRows: any[]) => {
      selectedRows.value = nextSelectedRows;
    };

    const setSelectedRowKeys = (nextSelectedRowKeys: Key[]) => {
      selectedRowKeys.value = nextSelectedRowKeys;
    };

    watchEffect(() => {
      setValidateLanguage(props.language ?? 'zh-CN');
    });

    const scrollY = ref(450);

    const { getRecordKey } = useItemKey(true);

    const tableRowKey = (record: any) => {
      const { rowKey } = props;

      if (typeof rowKey === 'function') {
        return rowKey(record);
      }

      return record[rowKey] || getRecordKey(record);
    };

    watchEffect(() => {
      if (contentRect.value) {
        const { height } = contentRect.value;

        scrollY.value = height - 140;
      }
    });

    watchEffect(() => {
      const form = formRef.value;
      const dataSource = dataSourceRef.value;
      const { hasClearSelectedRows } = props;

      if (form.id) {
        form.setValuesIn(LIST_FILED_CODE, dataSource);

        if (hasClearSelectedRows) {
          setSelectedRows([]);
          setSelectedRowKeys([]);
        }
      }
    });

    watchEffect(() => {
      const defaultSelectedRows = defaultSelectedRowsRef.value;

      setSelectedRowKeys(defaultSelectedRows?.map((item: any) => tableRowKey(item)) || []);
      setSelectedRows(defaultSelectedRows || []);
    });

    const handleRowSelectionChange = (rowKeys: Key[], changeSelectedRows: any[], info) => {
      const {
        rowSelectionType,
        rowSelection,
        selectedRows: defaultSelectedRows,
        dataSource,
      } = props;

      if (rowSelectionType === 'radio') {
        setSelectedRows(changeSelectedRows);
        setSelectedRowKeys(rowKeys);

        // @ts-ignore
        rowSelection?.onChange?.(rowKeys, changeSelectedRows, info);
      } else {
        const allDataSourceKey =
          toArray(dataSource).map((item: any, index) => tableRowKey(item) || index) || [];

        // 更新旧值选中的数据
        const selectedRowsMap = toArray(defaultSelectedRows).map((item: any) => {
          const record = changeSelectedRows.find(
            (current) => tableRowKey(current) === tableRowKey(item),
          );
          if (record) {
            return record;
          }
          return item;
        });

        const nextSelectedRows = selectedRowsMap
          .filter((item: Record<string, any>) => !allDataSourceKey.includes(tableRowKey(item)))
          .concat(
            changeSelectedRows.filter((item) => allDataSourceKey.includes(tableRowKey(item))),
          );

        const nextSelectedRowKeys = nextSelectedRows.map((item: any) => tableRowKey(item));

        setSelectedRows(nextSelectedRows);

        setSelectedRowKeys(nextSelectedRowKeys);
        // @ts-ignore
        rowSelection?.onChange?.(nextSelectedRowKeys, nextSelectedRows, info);
      }
    };

    const getRowSelectionConfig = computed(() => {
      const { hasRowSelection, rowSelection, rowSelectionType } = props;
      if (hasRowSelection) {
        return {
          // @ts-ignore
          ...rowSelection,
          selectedRowKeys: selectedRowKeys.value,
          selectedRows: selectedRows.value,
          type: rowSelectionType ?? 'checkbox',
          onChange: handleRowSelectionChange,
        };
      }
      return null;
    });

    const setComponentProps = () => {
      const form = formRef.value;
      const { onChange, pageSize, currentPage, total, tableLoading, ...restProps } = props;

      form.query(LIST_FILED_CODE).take((target) => {
        target.setComponentProps({
          ...restProps,
          loading: tableLoading,
          rowKey: tableRowKey,
          rowSelection: getRowSelectionConfig.value,
          onTableChange: onChange,
          selectedRowKeys: selectedRowKeys.value,
          setSelectedRowKeys,
          selectedRows: selectedRows.value,
          setSelectedRows,
          scrollY: scrollY.value,
          pagination: {
            pageSize: Number(pageSize ?? 30),
            current: Number(currentPage ?? 1),
            total: Number(total ?? 0),
          },
        });
      });
    };

    onMounted(() => {
      setComponentProps();
    });

    watch([getRowSelectionConfig, scrollY], () => {
      setComponentProps();
    });

    watch(
      () => {
        return {
          ...props,
        };
      },
      () => {
        setComponentProps();
      },
    );

    return () => {
      const { schema, form, loading } = props;

      return (
        <Skeleton loading={!!loading}>
          <FormProvider form={form}>
            <div
              class={`form-id-${form?.id}`}
              style={{
                height: '100%',
              }}
              ref={domRef}
            >
              {schema ? (
                <SchemaField schema={schema} components={components} />
              ) : (
                <Empty description="暂无数据" />
              )}
            </div>
          </FormProvider>
        </Skeleton>
      );
    };
  },
});

export default SchemeTableForm;
