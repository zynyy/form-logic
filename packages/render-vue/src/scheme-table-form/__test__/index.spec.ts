import { mount } from '@vue/test-utils';

import SchemeTableForm from '../SchemeTableForm';

import metaSchema from './metaSchema';
import { createForm } from '@formily/core';
import { TransformsSchema } from '@/transforms';

describe('schema table form ', () => {
  let wrapper;

  beforeEach(() => {
    const listSchema = new TransformsSchema({ metaSchema }).getListSchema();

    wrapper = mount(SchemeTableForm, {
      props: {
        schema: listSchema.tableSchema,
        form: createForm(),
      },
    });
  });

  test('gen snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('gen data snapshot', async () => {
    await wrapper.setProps({
      dataSource: [{ code: 'test snapshot', name: '测试生成快照' }],
      pageSize: 30,
      currentPage: 1,
      total: 20,
      rowKey: 'code',
    });

    await wrapper.vm.$forceUpdate();

    expect(wrapper.html()).toMatchSnapshot();
  });
});
