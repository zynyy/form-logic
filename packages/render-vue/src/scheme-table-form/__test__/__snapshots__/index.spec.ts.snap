// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema table form  gen data snapshot 1`] = `
<div class="form-id-snapshot" style="height: 100%;">
  <div class="form-logic-list-table">
    <div class="ant-table-wrapper">
      <div class="ant-spin-nested-loading">
        <!---->
        <div class="ant-spin-container">
          <div class="ant-table ant-table-fixed-header ant-table-fixed-column ant-table-scroll-horizontal ant-table-has-fix-left ant-table-has-fix-right ant-table-small ant-table-bordered">
            <div class="ant-table-title">
              <div class="ant-space ant-space-horizontal ant-space-align-center">
                <div class="ant-space-item"><button class="ant-btn ant-btn-primary" type="button"><span role="img" aria-label="plus" class="anticon anticon-plus"><svg focusable="false" class="" data-icon="plus" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><defs><style></style></defs><path d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"></path><path d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"></path></svg></span><span>新建</span></button></div>
                <!---->
              </div>
            </div>
            <div class="ant-table-container">
              <div style="overflow: hidden;" class="ant-table-header">
                <table style="table-layout: fixed; visibility: hidden;">
                  <colgroup></colgroup>
                  <thead class="ant-table-thead">
                    <tr>
                      <th title="序号" class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last ant-table-cell-ellipsis" style="text-align: center; position: sticky; left: 0px;" colstart="0" colend="0">
                        <!----><span class="ant-table-cell-content">序号</span>
                        <!---->
                      </th>
                      <th title="编码" class="ant-table-cell ant-table-cell-ellipsis" colstart="1" colend="1">
                        <!---->编码
                        <!---->
                      </th>
                      <th title="名称" class="ant-table-cell ant-table-cell-ellipsis" colstart="2" colend="2">
                        <!---->名称
                        <!---->
                      </th>
                      <th title="备注说明" class="ant-table-cell ant-table-cell-ellipsis" colstart="3" colend="3">
                        <!---->备注说明
                        <!---->
                      </th>
                      <th title="操作列" class="ant-table-cell ant-table-cell-fix-right ant-table-cell-fix-right-first ant-table-cell-ellipsis" style="position: sticky; right: 0px;" colstart="4" colend="4">
                        <!----><span class="ant-table-cell-content">操作列</span>
                        <!---->
                      </th>
                    </tr>
                  </thead>
                  <!---->
                </table>
              </div>
              <div style="overflow-x: auto; overflow-y: scroll; max-height: -140px;" class="ant-table-body">
                <table style="min-width: 100%; table-layout: auto;">
                  <colgroup>
                    <col style="width: 50px;">
                    <col style="width: 120px;">
                    <col style="width: 120px;">
                    <col style="width: 120px;">
                    <col style="width: 120px;">
                  </colgroup>
                  <tbody class="ant-table-tbody">
                    <tr aria-hidden="true" class="ant-table-measure-row" style="height: 0px; font-size: 0px;">
                      <td style="padding: 0px; border: 0px; height: 0px;">
                        <div style="height: 0px; overflow: hidden;">&nbsp;</div>
                      </td>
                      <td style="padding: 0px; border: 0px; height: 0px;">
                        <div style="height: 0px; overflow: hidden;">&nbsp;</div>
                      </td>
                      <td style="padding: 0px; border: 0px; height: 0px;">
                        <div style="height: 0px; overflow: hidden;">&nbsp;</div>
                      </td>
                      <td style="padding: 0px; border: 0px; height: 0px;">
                        <div style="height: 0px; overflow: hidden;">&nbsp;</div>
                      </td>
                      <td style="padding: 0px; border: 0px; height: 0px;">
                        <div style="height: 0px; overflow: hidden;">&nbsp;</div>
                      </td>
                    </tr>
                    <tr data-row-key="test snapshot" class="ant-table-row ant-table-row-level-0">
                      <td class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last ant-table-cell-ellipsis" style="text-align: center; position: sticky; left: 0px;">
                        <!----><span class="ant-table-cell-content"><span class="form-logic-array-base-index">#1</span></span>
                        <!---->
                      </td>
                      <td class="ant-table-cell ant-table-cell-ellipsis">
                        <!---->
                        <div style="grid-row-end: span 1; grid-column: span undefined / auto;" class="form-logic-form-item form-logic-form-item-fullness form-logic-form-item-label-align-right form-logic-form-item-control-align-left form-logic-form-item-control-wrap">
                          <!---->
                          <div class="form-logic-form-item-control form-logic-form-item-item-col-24">
                            <div class="form-logic-form-item-control-content">
                              <div class="form-logic-form-item-control-content-component">
                                <!----><span><div class="form-logic-array-base-preview-text" disabled="false" readonly="false">test snapshot</div></span>
                              </div>
                            </div>
                            <!---->
                          </div>
                        </div>
                        <!---->
                      </td>
                      <td class="ant-table-cell ant-table-cell-ellipsis">
                        <!---->
                        <div style="grid-row-end: span 1; grid-column: span undefined / auto;" class="form-logic-form-item form-logic-form-item-fullness form-logic-form-item-label-align-right form-logic-form-item-control-align-left form-logic-form-item-control-wrap">
                          <!---->
                          <div class="form-logic-form-item-control form-logic-form-item-item-col-24">
                            <div class="form-logic-form-item-control-content">
                              <div class="form-logic-form-item-control-content-component">
                                <!----><span><div class="form-logic-array-base-preview-text" disabled="false" readonly="false">测试生成快照</div></span>
                              </div>
                            </div>
                            <!---->
                          </div>
                        </div>
                        <!---->
                      </td>
                      <td class="ant-table-cell ant-table-cell-ellipsis">
                        <!---->
                        <div style="grid-row-end: span 1; grid-column: span undefined / auto;" class="form-logic-form-item form-logic-form-item-fullness form-logic-form-item-label-align-right form-logic-form-item-control-align-left form-logic-form-item-control-wrap">
                          <!---->
                          <div class="form-logic-form-item-control form-logic-form-item-item-col-24">
                            <div class="form-logic-form-item-control-content">
                              <div class="form-logic-form-item-control-content-component">
                                <!----><span><div class="form-logic-array-base-preview-text" disabled="false" readonly="false"><!----></div></span>
                              </div>
                            </div>
                            <!---->
                          </div>
                        </div>
                        <!---->
                      </td>
                      <td class="ant-table-cell ant-table-cell-fix-right ant-table-cell-fix-right-first ant-table-cell-ellipsis" style="position: sticky; right: 0px;">
                        <!----><span class="ant-table-cell-content"><div class="ant-space ant-space-horizontal ant-space-align-center"><div class="ant-space-item" style="margin-right: 8px;"><!----><button class="ant-btn ant-btn-text ant-btn-icon-only" type="button"><span role="img" aria-label="edit" class="anticon anticon-edit"><svg focusable="false" class="" data-icon="edit" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><path d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 000-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 009.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"></path></svg></span></button>
              </div>
              <!---->
              <div class="ant-space-item" style="margin-right: 8px;">
                <!----><button class="ant-btn ant-btn-text ant-btn-icon-only" type="button"><span role="img" aria-label="delete" class="anticon anticon-delete"><svg focusable="false" class="" data-icon="delete" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><path d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"></path></svg></span></button>
              </div>
              <!---->
              <div class="ant-space-item">
                <!----><button class="ant-btn ant-btn-text ant-btn-icon-only" type="button"><span role="img" aria-label="eye" class="anticon anticon-eye"><svg focusable="false" class="" data-icon="eye" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><path d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"></path></svg></span></button>
              </div>
              <!---->
            </div></span>
            <!---->
            </td>
            </tr>
            <!---->
            </tbody>
            <!---->
            </table>
          </div>
          <!---->
          <!---->
        </div>
        <!---->
      </div>
      <ul unselectable="on" class="ant-pagination mini ant-table-pagination ant-table-pagination-right">
        <li class="ant-pagination-total-text">总计20条; 当前页第1到20条</li>
        <li title="Previous Page" class="ant-pagination-prev ant-pagination-disabled" aria-disabled="true"><button class="ant-pagination-item-link" type="button" tabindex="-1" disabled=""><span role="img" aria-label="left" class="anticon anticon-left"><svg focusable="false" class="" data-icon="left" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><path d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"></path></svg></span></button></li>
        <li title="1" tabindex="0" class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"><a rel="nofollow">1</a></li>
        <li title="Next Page" class="ant-pagination-next ant-pagination-disabled" aria-disabled="true"><button class="ant-pagination-item-link" type="button" tabindex="-1" disabled=""><span role="img" aria-label="right" class="anticon anticon-right"><svg focusable="false" class="" data-icon="right" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><path d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"></path></svg></span></button></li>
        <li class="ant-pagination-options">
          <div class="ant-select ant-select-sm ant-pagination-options-size-changer ant-select-single ant-select-show-arrow">
            <!---->
            <!---->
            <div class="ant-select-selector"><span class="ant-select-selection-search"><input type="search" id="rc_select_0" autocomplete="off" class="ant-select-selection-search-input" role="combobox" aria-haspopup="listbox" aria-owns="rc_select_0_list" aria-autocomplete="list" aria-controls="rc_select_0_list" aria-activedescendant="rc_select_0_list_0" readonly="" unselectable="on" style="opacity: 0;"></span><span class="ant-select-selection-item">30 / page</span>
              <!---->
            </div><span class="ant-select-arrow" style="user-select: none;" unselectable="on" aria-hidden="true"><span role="img" aria-label="down" class="anticon anticon-down ant-select-suffix"><svg focusable="false" class="" data-icon="down" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><path d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"></path></svg></span></span>
            <!---->
          </div>
          <!---->
        </li>
      </ul>
    </div>
  </div>
</div>
<!---->
<!---->
<!---->
<!---->
<!---->
</div>
</div>
`;

exports[`schema table form  gen snapshot 1`] = `
<div class="form-id-snapshot" style="height: 100%;">
  <div class="form-logic-list-table">
    <div class="ant-table-wrapper">
      <div class="ant-spin-nested-loading">
        <!---->
        <div class="ant-spin-container">
          <div class="ant-table ant-table-fixed-header ant-table-fixed-column ant-table-scroll-horizontal ant-table-has-fix-left ant-table-has-fix-right ant-table-small ant-table-bordered ant-table-empty">
            <div class="ant-table-title">
              <div class="ant-space ant-space-horizontal ant-space-align-center">
                <div class="ant-space-item"><button class="ant-btn ant-btn-primary" type="button"><span role="img" aria-label="plus" class="anticon anticon-plus"><svg focusable="false" class="" data-icon="plus" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><defs><style></style></defs><path d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"></path><path d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"></path></svg></span><span>新建</span></button></div>
                <!---->
              </div>
            </div>
            <div class="ant-table-container">
              <div style="overflow: hidden;" class="ant-table-header">
                <table style="table-layout: fixed;">
                  <colgroup></colgroup>
                  <thead class="ant-table-thead">
                    <tr>
                      <th title="序号" class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last ant-table-cell-ellipsis" style="text-align: center; position: sticky; left: 0px;" colstart="0" colend="0">
                        <!----><span class="ant-table-cell-content">序号</span>
                        <!---->
                      </th>
                      <th title="编码" class="ant-table-cell ant-table-cell-ellipsis" colstart="1" colend="1">
                        <!---->编码
                        <!---->
                      </th>
                      <th title="名称" class="ant-table-cell ant-table-cell-ellipsis" colstart="2" colend="2">
                        <!---->名称
                        <!---->
                      </th>
                      <th title="备注说明" class="ant-table-cell ant-table-cell-ellipsis" colstart="3" colend="3">
                        <!---->备注说明
                        <!---->
                      </th>
                      <th title="操作列" class="ant-table-cell ant-table-cell-fix-right ant-table-cell-fix-right-first ant-table-cell-ellipsis" style="position: sticky; right: 0px;" colstart="4" colend="4">
                        <!----><span class="ant-table-cell-content">操作列</span>
                        <!---->
                      </th>
                    </tr>
                  </thead>
                  <!---->
                </table>
              </div>
              <div style="overflow-x: auto; overflow-y: scroll; max-height: -140px;" class="ant-table-body">
                <table style="min-width: 100%; table-layout: auto;">
                  <colgroup>
                    <col style="width: 50px;">
                    <col style="width: 120px;">
                    <col style="width: 120px;">
                    <col style="width: 120px;">
                    <col style="width: 120px;">
                  </colgroup>
                  <tbody class="ant-table-tbody">
                    <tr aria-hidden="true" class="ant-table-measure-row" style="height: 0px; font-size: 0px;">
                      <td style="padding: 0px; border: 0px; height: 0px;">
                        <div style="height: 0px; overflow: hidden;">&nbsp;</div>
                      </td>
                      <td style="padding: 0px; border: 0px; height: 0px;">
                        <div style="height: 0px; overflow: hidden;">&nbsp;</div>
                      </td>
                      <td style="padding: 0px; border: 0px; height: 0px;">
                        <div style="height: 0px; overflow: hidden;">&nbsp;</div>
                      </td>
                      <td style="padding: 0px; border: 0px; height: 0px;">
                        <div style="height: 0px; overflow: hidden;">&nbsp;</div>
                      </td>
                      <td style="padding: 0px; border: 0px; height: 0px;">
                        <div style="height: 0px; overflow: hidden;">&nbsp;</div>
                      </td>
                    </tr>
                    <tr class="ant-table-placeholder">
                      <td class="ant-table-cell" colspan="5">
                        <!---->
                        <div style="width: 0px; position: sticky; left: 0px; overflow: hidden;" class="ant-table-expanded-row-fixed">
                          <div class="ant-empty ant-empty-normal">
                            <div class="ant-empty-image"><svg class="ant-empty-img-simple" width="64" height="41" viewBox="0 0 64 41">
                                <g transform="translate(0 1)" fill="none" fill-rule="evenodd">
                                  <ellipse class="ant-empty-img-simple-ellipse" fill="#F5F5F5" cx="32" cy="33" rx="32" ry="7"></ellipse>
                                  <g class="ant-empty-img-simple-g" fill-rule="nonzero" stroke="#D9D9D9">
                                    <path d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"></path>
                                    <path d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z" fill="#FAFAFA" class="ant-empty-img-simple-path"></path>
                                  </g>
                                </g>
                              </svg></div>
                            <p class="ant-empty-description">No Data</p>
                            <!---->
                          </div>
                        </div>
                        <!---->
                      </td>
                    </tr>
                  </tbody>
                  <!---->
                </table>
              </div>
              <!---->
              <!---->
            </div>
            <!---->
          </div>
        </div>
      </div>
    </div>
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
  </div>
</div>
`;
